File: python/gitcc/commit_validator.py
from dataclasses import dataclass
from enum import Enum
from typing import Optional

from git.objects import Commit


class Status(str, Enum):
    """
    Status of a commit check.
    All none failures count as a success.
    """

    FAILURE = "FAILURE"
    WARNING = "WARNING"  # noqa: A003
    OK = "Correct"


@dataclass
class Result:
    """
    Result of a commit check. Commit might be omitted when only a message is checked.
    """

    status: Status = Status.FAILURE
    message: str = ""
    commit: Optional[Commit] = None

    def __str__(self) -> str:  # noqa: D105
        msg: str = self.status.value
        if self.commit is not None:
            msg = f"{msg} | {self.commit.hexsha} - {str(self.commit.summary)}"
        if self.message != "":
            if self.commit is None:
                msg = f"{msg} |"
            msg = f"{msg}\n        : {self.message}"
        return msg


def split_message(message: str) -> tuple[str, str]:
    """
    Split the commit message into summary and description.
    """
    res: list[str] = message.split('\n', 1)
    if len(res) == 1:
        return res[0], ""
    return res[0], res[1]


class CommitValidator:
    """
    Validator for commits.
    """

    def validate(self, commit: Commit) -> Result:
        """
        Validate a commit object.
        """
        res: Result = self.validate_message(*split_message(str(commit.message)))
        res.commit = commit
        return res

    def validate_message(self, _summary: str, _description: str) -> Result:
        """
        Check a commit message.
        """
        return Result(Status.OK, "")

File: typescript/src/commit-validator.ts
import {Commit} from './commit'

export enum Status {
    Failure = 'Failure',
    Warning = 'Warning',
    Ok = 'Correct'
}

export class Result {
    status: Status = Status.Failure
    message = ''
    commit: Commit | undefined = undefined

    constructor(status: Status, message = '', commit: Commit | undefined = undefined) {
        this.status = status
        this.message = message
        this.commit = commit
    }

    toString(colored = false): string {
        let msg: string = this.status.toString()
        if (colored) {
            switch (this.status) {
                case Status.Failure:
                    msg = `\u001b[0;31m${msg}\u001b[0m`
                    break
                case Status.Warning:
                    msg = `\u001b[0;33m${msg}\u001b[0m`
                    break
                case Status.Ok:
                    msg = `\u001b[0;32m${msg}\u001b[0m`
                    break
            }
        }
        if (this.commit !== undefined) {
            msg += ` | ${this.commit.hexsha} - ${this.commit.summary()}`
        }
        if (this.message !== '') {
            if (this.commit === undefined) {
                msg += ' |'
            }
            msg += `\n        : ${this.message}`
        }
        return msg
    }
}

export class CommitValidator {
    protected options: string[]
    constructor(options: string[]) {
        this.options = options
    }

    static split_message(message: string): [string, string] {
        const res: string[] = message.split('\n', 1)
        if (res.length === 1) {
            return [res[0], '']
        }
        return [res[0], res[1]]
    }

    validate(commit: Commit): Result {
        if (commit.message === undefined) {
            return new Result(Status.Failure, "commit message was empty.")
        }
        const check: Result = this.validate_message(...CommitValidator.split_message(commit.message))
        if (check.commit === undefined) {
            check.commit = commit
        }
        return check
    }

    validate_message(_summary: string, _description: string): Result {
        return new Result(Status.Ok, '')
    }
}
